packages/code-du-travail-frontend/src/lib/useLocalStorage.js

```diff
@@ -0,0 +1,50 @@
+import { useEffect, useState, useCallback } from "react";
+/**
+ * This localStorage hooks
+ * does not use a storage event handler by default
+ * since it double the render pass (setValue is called twice)
+ *
+ * @param {*} key the localStorage key
+ * @param {*} defaultValue the key's initialValue
+ */
+export function useLocalStorage(key, defaultValue, useStorageListener = false) {
+  const initialValue = () => {
+    try {
+      const data = window.localStorage.getItem(key);
+      return data !== null ? JSON.parse(data) : defaultValue;
+    } catch (error) {
+      return defaultValue;
+    }
+  };
+
+  const [value, setValue] = useState(initialValue);
+  const onStorage = useCallback(
+    event => {
+      if (event.storageArea === localStorage && event.key === key) {
+        try {
+          const data = JSON.parse(event.newValue);
+          setValue(data);
+        } catch (error) {
+          setValue(event.newValue);
+        }
+      }
+    },
+    [key, setValue]
+  );
+
+  useEffect(() => {
+    window.localStorage.setItem(key, JSON.stringify(value));
+    if (useStorageListener) {
+      window.addEventListener("storage", onStorage);
+      return () => {
+        window.removeEventListener("storage", onStorage);
+      };
+    }
+
+    // INFO(@lionelb): we use the result of JSON.stringify as
+    // dependency so IE11 doesn't loop endlessly
+    // eslint-disable-next-line
+  }, [key, JSON.stringify(value)]);
```

Comment:
Je suis étonné que ça fonctionne avec JSON.stringify(value) vu que du coup la valeur risque de changer même quand l'objet est le même non ? Je ne connais pas les mécanismes "under the hook" (pun intended) pour comparer `value`donc je n'en sais rien mais il y a un risque

Reply:
non justement,  c'est l'inverse. la reférence a l'objet peut changer, mais si le contenu change pas, on change rien